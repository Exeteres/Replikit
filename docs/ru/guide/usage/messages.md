# Сообщения

Пакет `@replikit/messages` содержит весь необходимый инструментарий для работы с сообщениями.

## MessageBuilder

Самый мощный интерфейс в Replikit для создания сообщений; использует паттерн `Fluent Builder`.

```ts
import { MessageBuilder } from "@replikit/messages";
import { AttachmentType } from "@replikit/core";

// Простое сообщение с текстом
const message1 = new MessageBuilder()
    .addText("Hello")
    .build();

// Сообщение, состоящее из нескольких строк
const message2 = new MessageBuilder()
    .addLine("one")
    .addLine("two")
    .addLines(["three", "four"])
    .build();

// Сообщение, содержащее вложение
const message3 = new MessageBuilder()
    .addAttachmentByUrl(AttachmentType.Photo, "https://example.com/photo.png")
    .build();
```

## Сокращения

Для создания простых сообщений существуют функции - сокращения:

```ts
import { fromText, fromCode, fromAttachment } from "@replikit/messages";
import { AttachmentType } from "@replikit/core";

// Простое сообщение с текстом
const message1 = fromText("Hello");

// Простое сообщение с кодом
const message2 = fromCode("Hello");

// Сообщение, содержащее вложение
const message3 = fromAttachment({
    type: AttachmentType.Photo,
    url: "https://example.com/photo.png"
});
```
