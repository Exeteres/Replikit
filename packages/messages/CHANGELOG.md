# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.22.1](https://github.com/Exeteres/Replikit/compare/v0.22.0...v0.22.1) (2020-10-23)

**Note:** Version bump only for package @replikit/messages





# [0.22.0](https://github.com/Exeteres/Replikit/compare/v0.21.0...v0.22.0) (2020-10-23)

**Note:** Version bump only for package @replikit/messages





# [0.21.0](https://github.com/Exeteres/Replikit/compare/v0.20.0...v0.21.0) (2020-10-17)

**Note:** Version bump only for package @replikit/messages





# [0.20.0](https://github.com/Exeteres/Replikit/compare/v0.19.0...v0.20.0) (2020-09-28)


### Features

* Add ability to create multiple button rows ([2a83a74](https://github.com/Exeteres/Replikit/commit/2a83a7432c102ff77e71cd4d542fdb4ec99ac7ad))





# [0.19.0](https://github.com/Exeteres/Replikit/compare/v0.18.2...v0.19.0) (2020-09-18)


### Features

* **messages:** Add addButton method ([1d7f3a0](https://github.com/Exeteres/Replikit/commit/1d7f3a0adb2b31e02707f4093e0ec7d828797403))
* **messages:** Add OutMessageLike type and utils ([a9fb59c](https://github.com/Exeteres/Replikit/commit/a9fb59cb432419fadd3a51e79d135ac29c9087b5))
* **messages:** Add token factories and support for buttons ([2275483](https://github.com/Exeteres/Replikit/commit/227548389a70cf64756c298951241ea96fbd77c4))





## [0.18.2](https://github.com/Exeteres/Replikit/compare/v0.18.1...v0.18.2) (2020-09-09)

**Note:** Version bump only for package @replikit/messages





## [0.18.1](https://github.com/Exeteres/Replikit/compare/v0.18.0...v0.18.1) (2020-09-09)

**Note:** Version bump only for package @replikit/messages





# [0.18.0](https://github.com/Exeteres/Replikit/compare/v0.17.3...v0.18.0) (2020-09-09)

**Note:** Version bump only for package @replikit/messages





## [0.17.3](https://github.com/Exeteres/Replikit/compare/v0.17.2...v0.17.3) (2020-08-25)

**Note:** Version bump only for package @replikit/messages





## [0.17.2](https://github.com/Exeteres/Replikit/compare/v0.17.1...v0.17.2) (2020-08-24)

**Note:** Version bump only for package @replikit/messages





## [0.17.1](https://github.com/Exeteres/Replikit/compare/v0.17.0...v0.17.1) (2020-08-21)

**Note:** Version bump only for package @replikit/messages





# [0.17.0](https://github.com/Exeteres/Replikit/compare/v0.16.0...v0.17.0) (2020-08-21)

**Note:** Version bump only for package @replikit/messages





# [0.16.0](https://github.com/Exeteres/Replikit/compare/v0.15.1...v0.16.0) (2020-08-18)


### Features

* **messages:** Add methods to work with message headers ([f412a28](https://github.com/Exeteres/Replikit/commit/f412a280c4a185a8066fec2b8492bc4afd8d8431))





## [0.15.1](https://github.com/Exeteres/Replikit/compare/v0.15.0...v0.15.1) (2020-08-13)

**Note:** Version bump only for package @replikit/messages





# [0.15.0](https://github.com/Exeteres/Replikit/compare/v0.14.0...v0.15.0) (2020-08-13)

**Note:** Version bump only for package @replikit/messages





# [0.14.0](https://github.com/Exeteres/Replikit/compare/v0.13.0...v0.14.0) (2020-08-05)


### Features

* **messages:** Add ability to use addLine method without text ([3886f11](https://github.com/Exeteres/Replikit/commit/3886f113aa94943afea2bd5f4f94462eccec055e))





# [0.13.0](https://github.com/Exeteres/Replikit/compare/v0.12.1...v0.13.0) (2020-08-04)


### Features

* **messages:** Add ability to use pipe method in MessageBuilder ([f5fe039](https://github.com/Exeteres/Replikit/commit/f5fe0391708753b7404e6c695a77cd793e83120c))
* **messages:** Add alias to use message with metadata as metadata ([e315a25](https://github.com/Exeteres/Replikit/commit/e315a256dfeef85137f640e08176d14c84a1638f))





# [0.12.0](https://github.com/Exeteres/Replikit/compare/v0.11.2...v0.12.0) (2020-07-30)

**Note:** Version bump only for package @replikit/messages





## [0.11.2](https://github.com/Exeteres/Replikit/compare/v0.11.1...v0.11.2) (2020-07-24)

**Note:** Version bump only for package @replikit/messages





## [0.11.1](https://github.com/Exeteres/Replikit/compare/v0.11.0...v0.11.1) (2020-07-24)

**Note:** Version bump only for package @replikit/messages






# [0.11.0](https://github.com/Exeteres/Replikit/compare/v0.10.1...v0.11.0) (2020-07-24)


### Features

* **messages:** Add MessageLike type ([046ca1b](https://github.com/Exeteres/Replikit/commit/046ca1b8a19308790f3e0a424a0cb9f3cb7dc7dd))





## [0.10.1](https://github.com/Exeteres/Replikit/compare/v0.10.0...v0.10.1) (2020-06-30)

**Note:** Version bump only for package @replikit/messages






# [0.10.0](https://github.com/Exeteres/Replikit/compare/v0.9.0...v0.10.0) (2020-06-29)

**Note:** Version bump only for package @replikit/messages





# [0.9.0](https://github.com/Exeteres/Replikit/compare/v0.8.0...v0.9.0) (2020-06-28)


### Features

* Add advanced metadata processing ([54c7820](https://github.com/Exeteres/Replikit/commit/54c782021f5ebc16d784f210d273f8d7490d2af5))





# [0.8.0](https://github.com/Exeteres/Replikit/compare/v0.7.2...v0.8.0) (2020-06-27)


### Features

* **messages:** Add MessageBuilder methods for working with replies ([015e3f2](https://github.com/Exeteres/Replikit/commit/015e3f23b3f166b46aecb2b98f6f447661b8576d))





## [0.7.2](https://github.com/Exeteres/Replikit/compare/v0.7.1...v0.7.2) (2020-06-23)

**Note:** Version bump only for package @replikit/messages





# [0.7.0](https://github.com/Exeteres/Replikit/compare/v0.6.0...v0.7.0) (2020-06-23)

**Note:** Version bump only for package @replikit/messages





# [0.6.0](https://github.com/Exeteres/Replikit/compare/v0.5.3...v0.6.0) (2020-06-21)

**Note:** Version bump only for package @replikit/messages





## [0.5.3](https://github.com/Exeteres/Replikit/compare/v0.5.2...v0.5.3) (2020-06-20)

**Note:** Version bump only for package @replikit/messages





## [0.5.2](https://github.com/Exeteres/Replikit/compare/v0.5.1...v0.5.2) (2020-06-20)

**Note:** Version bump only for package @replikit/messages





# 0.5.0 (2020-06-20)

**Note:** Version bump only for package @replikit/messages






# [0.4.0](https://github.com/Exeteres/Replikit/compare/v0.3.0...v0.4.0) (2020-06-19)

**Note:** Version bump only for package @replikit/messages





# [0.3.0](https://github.com/Exeteres/Replikit/compare/v0.2.0...v0.3.0) (2020-06-14)

**Note:** Version bump only for package @replikit/messages





# [0.2.0](https://github.com/Exeteres/Replikit/compare/v0.1.0...v0.2.0) (2020-06-13)

**Note:** Version bump only for package @replikit/messages
