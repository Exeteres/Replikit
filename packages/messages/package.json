{
    "name": "@replikit/messages",
    "version": "0.22.1",
    "license": "MIT",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "typings"
    ],
    "scripts": {
        "build": "tsc -p tsconfig.build.json"
    },
    "dependencies": {
        "@replikit/core": "^0.22.1"
    }
}
