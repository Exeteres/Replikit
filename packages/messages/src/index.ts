export * from "./shortcuts";
export * from "./utils";
export * from "./messageBuilder";
