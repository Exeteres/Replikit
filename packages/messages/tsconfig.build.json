{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "dist",
        "paths": {
            "@replikit/messages": ["messages/src"]
        }
    },
    "include": ["src"]
}
