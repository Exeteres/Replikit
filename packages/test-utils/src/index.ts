export * from "./databaseTestManager";
export * from "./tmp";
export * from "./code";
export * from "./expect";
export * from "./testController";
export * from "./commandTestManager";
export * from "./factories";
