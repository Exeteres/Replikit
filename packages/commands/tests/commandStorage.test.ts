import { createTestManager } from "@replikit/test-utils";
import { fromText, MessageBuilder } from "@replikit/messages";

describe("CommandStorage", () => {
    it.each([
        "/test abc 123",
        "/test abc",
        "/test abc abc",
        '/test "abc abc"',
        `/test "abc abc" 123`,
        `/test abc "123"`,
        "/test"
    ])(
        "should handle a command with required and optional parameters: %s",
        async text => {
            const { testManager, command } = createTestManager();
            command("test")
                .required("p1", String)
                .optional("p2", Number)
                .handler(context => {
                    expect({
                        p1: context.params.p1,
                        p2: context.params.p2
                    }).toMatchSnapshot();
                })
                .register();
            await testManager.processCommand(text);
            expect.assertions(1);
        }
    );

    it("should handle a command with alias", async () => {
        const { testManager, command } = createTestManager();
        command("test", "t")
            .handler(context => {
                expect(context.message.text).toBe("/t");
            })
            .register();
        await testManager.processCommand("/t");
        expect.assertions(1);
    });

    it.each(["/calc sum 20 22", "/calc diff 10 12", "/calc multiply"])(
        "should handle a command with subcommands: %s",
        async text => {
            const { testManager, command } = createTestManager();
            command("calc")
                .commands(
                    command("sum")
                        .required("p1", Number)
                        .required("p2", Number)
                        .handler(context => {
                            const result =
                                context.params.p1 + context.params.p2;
                            expect(result).toMatchSnapshot();
                        }),
                    command("diff")
                        .required("p1", Number)
                        .required("p2", Number)
                        .handler(context => {
                            const result =
                                context.params.p1 - context.params.p2;
                            expect(result).toMatchSnapshot();
                        })
                )
                .register();
            await testManager.processCommand(text);
            expect.assertions(1);
        }
    );

    it.each([
        ["a non-existing top-level command", "/test"],
        ["an undefined text", undefined],
        ["a non-command message", "hello"]
    ])("should ignore %s", async (_, text) => {
        const { testManager } = createTestManager();
        await testManager.processCommand(text!);
        expect.assertions(0);
    });

    it("should handle a command with a custom validator", async () => {
        const { testManager } = createTestManager();
        testManager
            .converter(Number)
            .validator((context, param, options) => {
                expect(param).toBe("123");
                expect(options.max).toBe(456);
                return +param + options.max!;
            })
            .register();
        testManager
            .command("test")
            .required("p1", Number, { max: 456 })
            .handler(context => {
                expect(context.params.p1).toBe(579);
            })
            .register();
        await testManager.processCommand("/test 123");
        expect.assertions(3);
    });

    it("should handle a command with a custom resolver", async () => {
        const { testManager } = createTestManager();
        testManager
            .converter(Number)
            .resolver((context, param, options) => {
                expect(param).toBe("123");
                expect(options.max).toBe(456);
                return Promise.resolve(+param + options.max!);
            })
            .register();
        testManager
            .command("test")
            .required("p1", Number, { max: 456 })
            .handler(context => {
                expect(context.params.p1).toBe(579);
            })
            .register();
        await testManager.processCommand("/test 123");
        expect.assertions(3);
    });

    it("should handle a command with a complete custom converter", async () => {
        const { testManager } = createTestManager();
        testManager
            .converter(Number)
            .validator((context, param) => [param + "0"])
            .resolver((context, param) => Promise.resolve(+param[0]))
            .register();
        testManager
            .command("test")
            .required("p1", Number)
            .handler(context => {
                expect(context.params.p1).toBe(20);
            })
            .register();
        await testManager.processCommand("/test 2");
        expect.assertions(1);
    });

    it.each(["validator", "resolver"] as const)(
        "should reply an error if %s returns a string",
        async method => {
            const { testManager } = createTestManager();
            testManager
                .converter(Number)
                // eslint-disable-next-line no-unexpected-multiline
                [method](() => "Invalid parameter")
                .register();
            testManager
                .command("test")
                .required("p1", Number)
                .handler(() => void 0)
                .register();
            await testManager.processCommand("/test 123");
            expect.assertions(1);
        }
    );

    it("should use a custom prefix", async () => {
        const { testManager } = createTestManager();
        testManager.commands.prefix = "!";
        testManager
            .command("test")
            .handler(context => expect(context.message.text).toBe("!test"))
            .register();
        await testManager.processCommand("!test");
        expect.assertions(1);
    });

    it("should handle an overloaded command", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .required("p1", String)
            .handler(context => {
                expect(context.message.text).toBe("/test one");
            })
            .register();
        command("test")
            .required("p1", String)
            .required("p2", String)
            .handler(context => {
                expect(context.message.text).toBe("/test one two");
            })
            .register();
        await testManager.processCommand("/test one");
        await testManager.processCommand("/test one two");
        expect.assertions(2);
    });

    it("should handle a nested overloaded command", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .commands(
                command("test")
                    .required("p1", String)
                    .handler(context => {
                        expect(context.message.text).toBe("/test test one");
                    }),
                command("test")
                    .required("p1", String)
                    .required("p2", String)
                    .handler(context => {
                        expect(context.message.text).toBe("/test test one two");
                    })
            )
            .register();
        await testManager.processCommand("/test test one");
        await testManager.processCommand("/test test one two");
        expect.assertions(2);
    });

    it.each([
        ["/test\ntest\ntest", "test\ntest"],
        ["/test test\ntest", "test\ntest"],
        ["/test test", "test"]
    ])(
        "should handle a command with text parameter: %s",
        async (text, result) => {
            const { testManager, command } = createTestManager();
            command("test")
                .text(false, true)
                .handler(context => {
                    expect(context.params.text).toBe(result);
                })
                .register();
            await testManager.processCommand(text);
            expect.assertions(1);
        }
    );

    it("should handle a command with text parameter with custom name", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .text("test")
            .handler(context => {
                expect(context.params.test).toBe("test\ntest");
            })
            .register();
        await testManager.processCommand("/test\ntest\ntest");
        expect.assertions(1);
    });

    it("should handle a command with text parameter splitted into lines", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .text(true)
            .handler(context => {
                const param = context.params.text;
                expect(param).toStrictEqual(["test", "test"]);
            })
            .register();
        await testManager.processCommand("/test\ntest\ntest");
        expect.assertions(1);
    });

    it("should handle a command with text parameter and skip validation", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .text(false, true)
            .handler(context => {
                expect(context.params.text).toBe("");
            })
            .register();
        await testManager.processCommand("/test");
        expect.assertions(1);
    });

    it("should handle a command with text parameter and reply an error", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .text()
            .handler(() => void 0)
            .register();
        await testManager.processCommand("/test");
        expect.assertions(1);
    });

    it("should handle a command with rest and text parameters", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .rest("strings", String)
            .text()
            .handler(context => {
                expect(context.params.strings).toMatchSnapshot();
                expect(context.params.text).toMatchSnapshot();
            })
            .register();
        await testManager.processCommand("/test abc 123\nfirst\nsecond");
        expect.assertions(2);
    });

    it.each(["/test 123", "/test 12 12", "/test", "/test abc"])(
        "should handle a command with rest parameters: %s",
        async text => {
            const { testManager, command } = createTestManager();
            command("test")
                .rest("numbers", Number)
                .handler(context => {
                    expect(context.params.numbers).toMatchSnapshot();
                })
                .register();
            await testManager.processCommand(text);
            expect.assertions(1);
        }
    );

    it("should handle a command with string rest parameters", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .rest("strings", String)
            .handler(context => {
                expect(context.params.strings).toMatchSnapshot();
            })
            .register();
        await testManager.processCommand("/test abc 123");
        expect.assertions(1);
    });

    it("should handle a command with rest parameters and resolver", async () => {
        const { testManager, command, converter } = createTestManager();
        converter(Number)
            .resolver((context, param) => +param)
            .register();
        command("test")
            .rest("numbers", Number)
            .handler(context => {
                expect(context.params.numbers).toMatchSnapshot();
            })
            .register();
        await testManager.processCommand("/test 123 456");
        expect.assertions(1);
    });

    it("should handle a command with rest parameters and resolver and reply an error", async () => {
        const { testManager, command, converter } = createTestManager();
        converter(Number)
            .resolver(() => "Invalid parameter")
            .register();
        command("test")
            .rest("numbers", Number)
            .handler(context => {
                expect(context.params.numbers).toMatchSnapshot();
            })
            .register();
        await testManager.processCommand("/test 123 456");
        expect.assertions(1);
    });

    it.each(["/test", "/test 123", "/test abc", "/test abc abc"])(
        "should handle a command with subcommands and default handler: %s",
        async text => {
            const { testManager, command } = createTestManager();
            command("test")
                .default("abc")
                .commands(
                    command("abc")
                        .required("p1", String)
                        .handler(context => {
                            expect(context.params.p1).toMatchSnapshot();
                        })
                )
                .register();
            await testManager.processCommand(text);
            expect.assertions(1);
        }
    );

    it("should handle a command returning OutMessage", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .handler(() => fromText("Hello"))
            .register();
        await testManager.processCommand("/test");
        expect.assertions(1);
    });

    it("should handle a command returning MessageBuilder", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .handler(() => new MessageBuilder().addText("Hello"))
            .register();
        await testManager.processCommand("/test");
        expect.assertions(1);
    });

    it("should handle a command with optional parameter with default value", async () => {
        const { testManager, command } = createTestManager();
        command("test")
            .optional("p1", Number, { default: 123 })
            .handler(context => {
                expect(context.params.p1).toBe(123);
            })
            .register();
        await testManager.processCommand("/test");
        expect.assertions(1);
    });

    it.each(["/test", "/test 1 2", "/test 1 2 3"])(
        "should handle a command with rest parameter count restriction: %s",
        async text => {
            const { testManager, command } = createTestManager();
            command("test")
                .rest("numbers", Number, { minCount: 1, maxCount: 2 })
                .handler(context => {
                    expect(context.params.numbers).toMatchSnapshot();
                })
                .register();
            await testManager.processCommand(text);
            expect.assertions(1);
        }
    );
});
