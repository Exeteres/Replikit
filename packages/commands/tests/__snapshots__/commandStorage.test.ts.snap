// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandStorage should handle a command returning MessageBuilder 1`] = `"Hello"`;

exports[`CommandStorage should handle a command returning OutMessage 1`] = `"Hello"`;

exports[`CommandStorage should handle a command with required and optional parameters: /test "abc abc" 1`] = `
Object {
  "p1": "abc abc",
  "p2": undefined,
}
`;

exports[`CommandStorage should handle a command with required and optional parameters: /test "abc abc" 123 1`] = `
Object {
  "p1": "abc abc",
  "p2": 123,
}
`;

exports[`CommandStorage should handle a command with required and optional parameters: /test 1`] = `
"Argument count mismatch.
1 expected, got 0.
Usage: /test {p1} [p2]"
`;

exports[`CommandStorage should handle a command with required and optional parameters: /test abc "123" 1`] = `
Object {
  "p1": "abc",
  "p2": 123,
}
`;

exports[`CommandStorage should handle a command with required and optional parameters: /test abc 1`] = `
Object {
  "p1": "abc",
  "p2": undefined,
}
`;

exports[`CommandStorage should handle a command with required and optional parameters: /test abc 123 1`] = `
Object {
  "p1": "abc",
  "p2": 123,
}
`;

exports[`CommandStorage should handle a command with required and optional parameters: /test abc abc 1`] = `
"[p2: Number required]
Usage: /test {p1} [p2]"
`;

exports[`CommandStorage should handle a command with rest and text parameters 1`] = `
Array [
  "abc",
  "123",
]
`;

exports[`CommandStorage should handle a command with rest and text parameters 2`] = `
"first
second"
`;

exports[`CommandStorage should handle a command with rest parameter count restriction: /test 1`] = `
"[numbers: The number of additional parameters should be no less than 1]
Usage: /test ...numbers"
`;

exports[`CommandStorage should handle a command with rest parameter count restriction: /test 1 2 1`] = `
Array [
  1,
  2,
]
`;

exports[`CommandStorage should handle a command with rest parameter count restriction: /test 1 2 3 1`] = `
"[numbers: The number of additional parameters should be no more than 2]
Usage: /test ...numbers"
`;

exports[`CommandStorage should handle a command with rest parameters and resolver 1`] = `
Array [
  123,
  456,
]
`;

exports[`CommandStorage should handle a command with rest parameters and resolver and reply an error 1`] = `
"[numbers: Invalid parameter]
Usage: /test ...numbers"
`;

exports[`CommandStorage should handle a command with rest parameters: /test 1`] = `Array []`;

exports[`CommandStorage should handle a command with rest parameters: /test 12 12 1`] = `
Array [
  12,
  12,
]
`;

exports[`CommandStorage should handle a command with rest parameters: /test 123 1`] = `
Array [
  123,
]
`;

exports[`CommandStorage should handle a command with rest parameters: /test abc 1`] = `
"[numbers: Number required]
Usage: /test ...numbers"
`;

exports[`CommandStorage should handle a command with string rest parameters 1`] = `
Array [
  "abc",
  "123",
]
`;

exports[`CommandStorage should handle a command with subcommands and default handler: /test 1`] = `
"Argument count mismatch.
1 expected, got 0.
Usage: /test abc {p1}"
`;

exports[`CommandStorage should handle a command with subcommands and default handler: /test 123 1`] = `"123"`;

exports[`CommandStorage should handle a command with subcommands and default handler: /test abc 1`] = `
"Argument count mismatch.
1 expected, got 0.
Usage: /test abc {p1}"
`;

exports[`CommandStorage should handle a command with subcommands and default handler: /test abc abc 1`] = `"abc"`;

exports[`CommandStorage should handle a command with subcommands: /calc diff 10 12 1`] = `-2`;

exports[`CommandStorage should handle a command with subcommands: /calc multiply 1`] = `
"Command not found
Usage: /calc {sum|diff}"
`;

exports[`CommandStorage should handle a command with subcommands: /calc sum 20 22 1`] = `42`;

exports[`CommandStorage should handle a command with text parameter and reply an error 1`] = `
"[text: Empty text parameter]
Usage: /test"
`;

exports[`CommandStorage should handle a command with text parameter: /test 1`] = `
"[text: Empty text parameter]
Usage: /test"
`;

exports[`CommandStorage should reply an error if default resolver returns empty array 1`] = `
"[p1: Error when resolving default value]
Usage: /test [p1]"
`;

exports[`CommandStorage should reply an error if default resolver returns undefined 1`] = `
"[p1: Error when resolving default value]
Usage: /test [p1]"
`;

exports[`CommandStorage should reply an error if resolver returns a string 1`] = `
"[p1: Invalid parameter]
Usage: /test {p1}"
`;

exports[`CommandStorage should reply an error if validator returns a string 1`] = `
"[p1: Invalid parameter]
Usage: /test {p1}"
`;

exports[`CommandStorage should reply an error returned from default resolver 1`] = `
"[p1: Invalid parameter]
Usage: /test [p1]"
`;
